File: pkg/client/request.go
package client

import (
	"encoding/binary"
	"nmid-v2/pkg/model"
	"nmid-v2/pkg/utils"
)

type Request struct {
	DataType uint32
	Data     []byte
	DataLen  uint32

	Handle           string
	HandleLen        uint32
	ParamsType       uint32
	ParamsHandleType uint32
	ParamsLen        uint32
	Params           []byte
	Ret              []byte
	RetLen           uint32
}

func NewReq() (req *Request) {
	req = &Request{
		Data:             make([]byte, 0),
		ParamsType:       model.PARAMS_TYPE_MSGPACK,
		ParamsHandleType: model.PARAMS_HANDLE_TYPE_ENCODE,
		Ret:              make([]byte, 0),
	}
	return
}

//ContentPack 打包内容
func (req *Request) ContentPack(dataType uint32, handle string, params []byte) (content []byte, contentLen uint32) {
	req.DataType = dataType
	req.Handle = handle
	req.HandleLen = uint32(len(handle))
	req.Params = params
	req.ParamsLen = uint32(len(params))
	req.DataLen = uint32(model.UINT32_SIZE + model.UINT32_SIZE + model.UINT32_SIZE + req.HandleLen + model.UINT32_SIZE + req.ParamsLen)
	contentLen = req.DataLen

	content = make([]byte, contentLen)
	binary.BigEndian.PutUint32(content[:model.UINT32_SIZE], req.ParamsType)
	start := model.UINT32_SIZE
	end := start + model.UINT32_SIZE
	binary.BigEndian.PutUint32(content[start:end], req.ParamsHandleType)
	start = end
	end = start + model.UINT32_SIZE
	binary.BigEndian.PutUint32(content[start:end], req.HandleLen)
	start = end
	end = start + int(req.HandleLen)
	copy(content[start:end], []byte(req.Handle))
	start = end
	end = start + model.UINT32_SIZE
	binary.BigEndian.PutUint32(content[start:end], req.ParamsLen)
	start = end
	end = start + int(req.ParamsLen)
	copy(content[start:end], req.Params)
	req.Data = content

	return
}

//EncodePack 打包
func (req *Request) EncodePack() (data []byte) {
	len := model.MIN_DATA_SIZE + req.DataLen //add 12 bytes head
	data = utils.GetBuffer(int(len))

	binary.BigEndian.PutUint32(data[:4], model.CONN_TYPE_CLIENT)
	binary.BigEndian.PutUint32(data[4:8], req.DataType)
	binary.BigEndian.PutUint32(data[8:model.MIN_DATA_SIZE], req.DataLen)
	copy(data[model.MIN_DATA_SIZE:], req.Data)

	return
}

File: sdk/java/nmidsdk/client/src/main/java/com/nmidsdk/client/Request.java
package com.nmidsdk.client;

import com.nmidsdk.model.Constants;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

/**
 * Client request encoder.
 * Mirrors sdk/node/lib/client/request.js.
 *
 * Packet layout (big-endian):
 *   header (12 bytes): ConnType(4) | DataType(4) | DataLen(4)
 *   body (DataLen bytes) for PDT_C_DO_JOB:
 *     ParamsType(4) | ParamsHandleType(4) | HandleLen(4) | Handle |
 *     ParamsLen(4) | Params
 */
public class Request {
    public int dataType;
    public byte[] data = new byte[0];
    public int dataLen;

    public String handle = "";
    public int handleLen;
    public int paramsType = Constants.PARAMS_TYPE_MSGPACK;
    public int paramsHandleType = Constants.PARAMS_HANDLE_TYPE_ENCODE;
    public int paramsLen;
    public byte[] params = new byte[0];

    /** Build the body for a client do-job request. */
    public void contentPack(int dataType, String handle, byte[] params) {
        this.dataType = dataType;
        this.handle = handle;
        byte[] handleBytes = handle.getBytes(StandardCharsets.UTF_8);
        this.handleLen = handleBytes.length;
        this.params = params == null ? new byte[0] : params;
        this.paramsLen = this.params.length;
        this.dataLen =
            Constants.UINT32_SIZE + // paramsType
            Constants.UINT32_SIZE + // paramsHandleType
            Constants.UINT32_SIZE + // handleLen
            this.handleLen +
            Constants.UINT32_SIZE + // paramsLen
            this.paramsLen;

        ByteBuffer content = ByteBuffer.allocate(this.dataLen);
        content.putInt(this.paramsType);
        content.putInt(this.paramsHandleType);
        content.putInt(this.handleLen);
        content.put(handleBytes);
        content.putInt(this.paramsLen);
        content.put(this.params);

        this.data = content.array();
    }

    /** Prepend the 12-byte header (CONN_TYPE_CLIENT). */
    public byte[] encodePack() {
        int total = Constants.MIN_DATA_SIZE + this.dataLen;
        ByteBuffer data = ByteBuffer.allocate(total);
        data.putInt(Constants.CONN_TYPE_CLIENT);
        data.putInt(this.dataType);
        data.putInt(this.dataLen);
        if (this.dataLen > 0) {
            data.put(this.data);
        }
        return data.array();
    }
}
